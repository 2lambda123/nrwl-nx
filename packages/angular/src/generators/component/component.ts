import { joinPathFragments, logger, names, readJson, Tree } from '@nrwl/devkit';
import type { Schema } from './schema';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import {
  formatFiles,
  readWorkspaceConfiguration,
  readProjectConfiguration,
  normalizePath,
} from '@nrwl/devkit';
import { normalize } from 'path';

export async function componentGenerator(tree: Tree, schema: Schema) {
  const { inlineScam, ...options } = schema;

  checkPathUnderProjectRoot(tree, options);

  const angularComponentSchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );
  await angularComponentSchematic(tree, {
    ...options,
    skipImport: true,
    export: false,
  });

  exportComponent(tree, schema);

  await formatFiles(tree);
}

function checkPathUnderProjectRoot(tree: Tree, options: Partial<Schema>) {
  if (!options.path) {
    return;
  }

  const project =
    options.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root } = readProjectConfiguration(tree, project);

  let pathToComponent = normalizePath(options.path);
  pathToComponent = pathToComponent.startsWith('/')
    ? pathToComponent.slice(1)
    : pathToComponent;

  if (!pathToComponent.startsWith(normalize(root))) {
    throw new Error(
      `The path provided for the component (${options.path}) does not exist under the project root (${root}).`
    );
  }
}

function exportComponent(tree: Tree, schema: Schema) {
  if (!schema.export) {
    return;
  }

  const project =
    schema.project ?? readWorkspaceConfiguration(tree).defaultProject;

  const { root, sourceRoot, projectType } = readProjectConfiguration(
    tree,
    project
  );

  if (projectType === 'application') {
    logger.warn(
      '--export=true was ignored as the project the component being generated in is not a library.'
    );

    return;
  }

  const componentNames = names(schema.name);

  const componentFileName = `${componentNames.fileName}.${
    schema.type ?? 'component'
  }`;

  let componentDirectory = schema.flat
    ? joinPathFragments(sourceRoot, 'lib')
    : joinPathFragments(sourceRoot, 'lib', componentNames.fileName);

  const componentFilePath = joinPathFragments(
    componentDirectory,
    `${componentFileName}.ts`
  );

  const ngPackageJsonPath = joinPathFragments(root, 'ng-package.json');
  const ngPackageEntryPoint = tree.exists(ngPackageJsonPath)
    ? readJson(tree, ngPackageJsonPath).lib?.entryFile
    : undefined;

  const projectEntryPoint = ngPackageEntryPoint
    ? joinPathFragments(root, ngPackageEntryPoint)
    : joinPathFragments(sourceRoot, `index.ts`);

  if (!tree.exists(projectEntryPoint)) {
    // Let's not error, simply warn the user
    // It's not too much effort to manually do this
    // It would be more frustrating to have to find the correct path and re-run the command
    logger.warn(
      `Could not export Componetn. Unable to determine project's entry point. Path ${projectEntryPoint} does not exist. Component has still been created.`
    );

    return;
  }

  const relativePathFromEntryPoint = `.${componentFilePath
    .split(sourceRoot)[1]
    .replace('.ts', '')}`;

  const updateEntryPointContent = `${tree.read(projectEntryPoint)}
    export * from "${relativePathFromEntryPoint}";`;

  tree.write(projectEntryPoint, updateEntryPointContent);
}

export default componentGenerator;
