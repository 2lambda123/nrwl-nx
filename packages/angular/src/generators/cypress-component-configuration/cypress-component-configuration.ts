import { cypressComponentProject } from '@nrwl/cypress';
import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { componentTestGenerator } from '../component-test/component-test';
import {
  getComponentsInfo,
  getStandaloneComponentsInfo,
} from '../utils/storybook-ast/component-info';
import { getProjectEntryPoints } from '../utils/storybook-ast/entry-point';
import { getModuleFilePaths } from '../utils/storybook-ast/module-info';
import { CypressComponentConfigSchema } from './schema';

/**
 * This is for cypress built in component testing, if you want to test with
 * storybook + cypress then use the componentCypressGenerator instead.
 */
export async function cypressComponentConfiguration(
  tree: Tree,
  options: CypressComponentConfigSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const installTask = await cypressComponentProject(tree, {
    project: options.project,
    skipFormat: true,
  });

  addFiles(tree, projectConfig, options);

  if (options.skipFormat) {
    await formatFiles(tree);
  }
  return () => {
    installTask();
  };
}
function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentConfigSchema
) {
  const cypressConfigPath = joinPathFragments(
    joinPathFragments(projectConfig.root, 'cypress.config.ts')
  );

  if (tree.exists(cypressConfigPath)) {
    tree.delete(cypressConfigPath);
  }
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      tpl: '',
    }
  );

  if (options.generateTests) {
    const entryPoints = getProjectEntryPoints(tree, options.project);

    const componentInfo = [];
    for (const entryPoint of entryPoints) {
      const moduleFilePaths = getModuleFilePaths(tree, entryPoint);
      componentInfo.push(
        ...getComponentsInfo(
          tree,
          entryPoint,
          moduleFilePaths,
          options.project
        ),
        ...getStandaloneComponentsInfo(tree, entryPoint)
      );
    }

    for (const info of componentInfo) {
      if (info === undefined) {
        continue;
      }
      componentTestGenerator(tree, {
        projectPath: info.moduleFolderPath,
        componentName: info.name,
        componentDir: info.path,
        componentFileName: info.componentFileName,
        skipFormat: true,
      });
    }
  }
}
