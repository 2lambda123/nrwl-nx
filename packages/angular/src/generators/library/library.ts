import {
  formatFiles,
  installPackagesTask,
  moveFilesToNewDirectory,
  Tree,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { jestProjectGenerator } from '@nrwl/jest';
import { Linter } from '@nrwl/workspace';

import init from '../../generators/init/init';
import addLintingGenerator from '../add-linting/add-linting';
import karmaProjectGenerator from '../karma-project/karma-project';

import { addModule } from './lib/add-module';
import { normalizeOptions } from './lib/normalize-options';
import { updateLibPackageNpmScope } from './lib/update-lib-package-npm-scope';
import { updateProject } from './lib/update-project';
import { updateTsConfig } from './lib/update-tsconfig';
import {
  enableStrictTypeChecking,
  setLibraryStrictDefault,
} from './lib/enable-strict-type-checking';
import { NormalizedSchema } from './lib/normalized-schema';
import { Schema } from './schema';

export default async function (host: Tree, schema: Schema) {
  // Do some validation checks
  const options = normalizeOptions(host, schema);
  if (!options.routing && options.lazy) {
    throw new Error(`routing must be set`);
  }

  if (options.enableIvy === true && !options.buildable) {
    throw new Error('enableIvy must only be used with buildable.');
  }

  if (options.publishable === true && !schema.importPath) {
    throw new Error(
      `For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`
    );
  }

  console.log('Initializing library generator');
  init(host, {
    ...options,
    skipFormat: true,
  });

  console.log('Scaffolding initial files');
  const runAngularLibrarySchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'library'
  );
  await runAngularLibrarySchematic(host, {
    name: options.name,
    prefix: options.prefix,
    entryFile: 'index',
    skipPackageJson: !(options.publishable || options.buildable),
    skipTsConfig: false,
  });

  console.log('Moving to correct directory');
  moveFilesToNewDirectory(host, options.name, options.projectRoot);

  console.log('Updating project configuration');
  await updateProject(host, options);

  console.log("Updating project's TS Config");
  updateTsConfig(host, options);

  console.log('Adding unit test runner');
  await addUnitTestRunner(host, options);

  console.log('If Buildable/Publishable, fix up npm scope');
  updateNpmScopeIfBuildableOrPublishable(host, options);

  console.log('Add Angular Module');
  addModule(host, options);

  console.log('Determine if strict mode should be set');
  setStrictMode(host, options);

  console.log('Add Linter');
  await addLinting(host, options);

  console.log('Formatting files');
  await formatFiles(host);

  return () => {
    installPackagesTask(host);
  };
}

async function addUnitTestRunner(host: Tree, options: NormalizedSchema) {
  if (options.unitTestRunner === 'jest') {
    await jestProjectGenerator(host, {
      project: options.name,
      setupFile: 'angular',
      supportTsx: false,
      skipSerializers: false,
    });
  } else if (options.unitTestRunner === 'karma') {
    await karmaProjectGenerator(host, {
      project: options.name,
    });
  }
}

function updateNpmScopeIfBuildableOrPublishable(
  host: Tree,
  options: NormalizedSchema
) {
  if (options.buildable || options.publishable) {
    updateLibPackageNpmScope(host, options);
  }
}

function setStrictMode(host: Tree, options: NormalizedSchema) {
  if (options.strict) {
    enableStrictTypeChecking(host, options);
  } else {
    setLibraryStrictDefault(host, options.strict);
  }
}

async function addLinting(host: Tree, options: NormalizedSchema) {
  if (options.linter === Linter.None) {
    return;
  }
  await addLintingGenerator(host, {
    projectName: options.name,
    projectRoot: options.projectRoot,
    prefix: options.prefix,
  });
}
