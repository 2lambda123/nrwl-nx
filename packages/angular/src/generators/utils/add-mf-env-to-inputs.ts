import { type Tree, readNxJson, updateNxJson } from '@nx/devkit';

export function addMfEnvToTargetDefaultInputs(tree: Tree) {
  const nxJson = readNxJson(tree);
  const webpackExecutor = '@nx/angular:webpack-browser';
  const mfEnvVar = 'NX_MF_DEV_SERVER_STATIC_REMOTES';

  let mfEnvVarExists = false;
  if (nxJson.targetDefaults && webpackExecutor in nxJson.targetDefaults) {
    const webpackExecutorTargetDefaults =
      nxJson.targetDefaults[webpackExecutor];

    if (webpackExecutorTargetDefaults.inputs) {
      for (const webpackExecutorTargetDefaultsInput of webpackExecutorTargetDefaults.inputs) {
        if (typeof webpackExecutorTargetDefaultsInput === 'object') {
          const [key, value] = Object.entries(
            webpackExecutorTargetDefaultsInput
          )[0];

          if (key === 'env' && value === mfEnvVar) {
            mfEnvVarExists = true;
            break;
          }
        }
      }
    } else {
      nxJson.targetDefaults[webpackExecutor].inputs = [];
    }
  } else if (!nxJson.targetDefaults[webpackExecutor]) {
    nxJson.targetDefaults[webpackExecutor] = {
      inputs: [],
    };
  } else if (!nxJson.targetDefaults) {
    nxJson.targetDefaults = {
      [webpackExecutor]: {
        inputs: [],
      },
    };
  }

  if (!mfEnvVarExists) {
    nxJson.targetDefaults[webpackExecutor].inputs.push({ env: mfEnvVar });
  }

  updateNxJson(tree, nxJson);
}
