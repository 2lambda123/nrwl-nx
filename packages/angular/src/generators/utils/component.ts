import {
  joinPathFragments,
  names,
  normalizePath,
  readProjectConfiguration,
  readWorkspaceConfiguration,
  Tree,
} from '@nrwl/devkit';

export type ComponentGenerationOptions = {
  name: string;
  flat?: boolean;
  path?: string;
  project?: string;
  type?: string;
};
export type ComponentFileInfo = {
  componentDirectory: string;
  componentFileName: string;
  componentFilePath: string;
};

export function getComponentFileInfo(
  tree: Tree,
  options: ComponentGenerationOptions
): ComponentFileInfo {
  const project =
    options.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root, sourceRoot, projectType } = readProjectConfiguration(
    tree,
    project
  );
  const { fileName: normalizedName } = names(options.name);

  const componentFileName = `${normalizedName}.${
    options.type ? names(options.type).fileName : 'component'
  }`;

  const projectSourceRoot = sourceRoot ?? joinPathFragments(root, 'src');
  const path =
    options.path ??
    joinPathFragments(
      projectSourceRoot,
      projectType === 'application' ? 'app' : 'lib'
    );
  const componentDirectory = options.flat
    ? normalizePath(path)
    : joinPathFragments(path, normalizedName);

  const componentFilePath = joinPathFragments(
    componentDirectory,
    `${componentFileName}.ts`
  );

  return { componentDirectory, componentFileName, componentFilePath };
}
