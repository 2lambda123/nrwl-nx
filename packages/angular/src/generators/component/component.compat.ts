import { convertNxGenerator } from '@nrwl/devkit';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import { componentGenerator } from './component';

export default warnForSchematicUsage(convertNxGenerator(componentGenerator));
