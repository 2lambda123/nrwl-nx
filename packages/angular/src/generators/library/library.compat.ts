import { convertNxGenerator } from '@nx/devkit';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import { libraryGeneratorInternal } from './library';

export const librarySchematic = warnForSchematicUsage(
  convertNxGenerator(libraryGeneratorInternal)
);
