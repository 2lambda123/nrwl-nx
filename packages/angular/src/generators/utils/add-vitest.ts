import {
  addDependenciesToPackageJson,
  ensurePackage,
  joinPathFragments,
  readJson,
  type Tree,
  updateJson,
} from '@nx/devkit';
import { analogVitestAngular, nxVersion } from '../../utils/versions';

export type AddVitestOptions = {
  name: string;
  projectRoot: string;
  skipPackageJson: boolean;
  strict: boolean;
};

export async function addVitest(
  tree: Tree,
  options: AddVitestOptions
): Promise<void> {
  if (!options.skipPackageJson) {
    process.env.npm_config_legacy_peer_deps ??= 'true';

    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@analogjs/vitest-angular': analogVitestAngular,
        '@analogjs/vite-plugin-angular': analogVitestAngular,
      },
      undefined,
      true
    );
  }

  const { createOrEditViteConfig, viteConfigurationGenerator } = ensurePackage<
    typeof import('@nx/vite')
  >('@nx/vite', nxVersion);

  const relativeTestSetupPath = joinPathFragments('src', 'test-setup.ts');
  const isEsmImplicit = readJson(tree, 'package.json').type === 'module';

  createOrEditViteConfig(
    tree,
    {
      project: options.name,
      includeLib: false,
      includeVitest: true,
      inSourceTests: false,
      imports: [`import angular from '@analogjs/vite-plugin-angular'`],
      plugins: ['angular()'],
      setupFile: relativeTestSetupPath,
      useEsmExtension: !isEsmImplicit,
    },
    true
  );

  const setupFile = joinPathFragments(
    options.projectRoot,
    relativeTestSetupPath
  );
  if (!tree.exists(setupFile)) {
    tree.write(
      setupFile,
      `import '@analogjs/vitest-angular/setup-zone';

import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting,
} from '@angular/platform-browser-dynamic/testing';
import { getTestBed } from '@angular/core/testing';

getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting()
);
`
    );
  }

  await viteConfigurationGenerator(tree, {
    project: options.name,
    newProject: true,
    uiFramework: 'angular',
    includeVitest: true,
    testEnvironment: 'jsdom',
  });

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.spec.json'),
    (tsConfig) => {
      tsConfig.files = [...(tsConfig.files ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    (tsConfig) => {
      tsConfig.exclude = [...(tsConfig.exclude ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );
}
