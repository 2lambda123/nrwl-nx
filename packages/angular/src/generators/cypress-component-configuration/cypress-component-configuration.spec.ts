import { installedCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import { joinPathFragments, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { componentGenerator } from '../component/component';
import librarySecondaryEntryPointGenerator from '../library-secondary-entry-point/library-secondary-entry-point';
import { libraryGenerator } from '../library/library';
import { cypressComponentConfiguration } from './cypress-component-configuration';

jest.mock('@nrwl/cypress/src/utils/cypress-version');
describe('Cypress Component Testing Configuration', () => {
  let tree: Tree;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    tree.write('.gitignore', '');
    mockedInstalledCypressVersion.mockReturnValue(10);
  });

  it('should work with simple components', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
    });

    await setup(tree, {
      project: 'my-lib',
      name: 'something',
      standalone: false,
    });

    await cypressComponentConfiguration(tree, {
      project: 'my-lib',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/my-lib/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with standalone component', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib-standalone',
    });

    await setup(tree, {
      project: 'my-lib-standalone',
      name: 'something',
      standalone: true,
    });

    await cypressComponentConfiguration(tree, {
      project: 'my-lib-standalone',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/my-lib-standalone/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with complex component', async () => {
    await libraryGenerator(tree, {
      name: 'with-inputs-cmp',
    });

    await setup(tree, {
      project: 'with-inputs-cmp',
      name: 'something',
      standalone: false,
      withInputs: true,
      basePath: 'libs/with-inputs-cmp/src/lib',
    });

    await cypressComponentConfiguration(tree, {
      project: 'with-inputs-cmp',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/with-inputs-cmp/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with complex standalone component', async () => {
    await libraryGenerator(tree, {
      name: 'with-inputs-standalone-cmp',
    });

    await setup(tree, {
      project: 'with-inputs-standalone-cmp',
      name: 'something',
      standalone: true,
      withInputs: true,
      basePath: 'libs/with-inputs-standalone-cmp/src/lib',
    });

    await cypressComponentConfiguration(tree, {
      project: 'with-inputs-standalone-cmp',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/with-inputs-standalone-cmp/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with secondary entry point libs', async () => {
    await libraryGenerator(tree, {
      name: 'secondary',
      buildable: true,
    });
    await librarySecondaryEntryPointGenerator(tree, {
      name: 'button',
      library: 'secondary',
    });
    await componentGenerator(tree, {
      name: 'fancy-button',
      path: 'libs/secondary/src/lib/button',
      project: 'secondary',
      flat: true,
    });

    await componentGenerator(tree, {
      name: 'standalone-fancy-button',
      path: 'libs/secondary/src/lib/button',
      project: 'secondary',
      standalone: true,
      flat: true,
    });

    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'secondary',
    });
    console.log(tree.listChanges().map((c) => c.path));
    expect(
      tree.read(
        'libs/secondary/src/lib/button/fancy-button.component.cy.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
    expect(
      tree.read(
        'libs/secondary/src/lib/button/standalone-fancy-button.component.cy.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
  });

  it('should not overwrite existing component test', async () => {
    await libraryGenerator(tree, {
      name: 'cool-lib',
      flat: true,
    });
    await setup(tree, { project: 'cool-lib', name: 'abc', standalone: false });
    tree.write(
      'libs/cool-lib/src/lib/abc-one/abc-one.component.cy.ts',
      'should not overwrite abc-one'
    );
    tree.write(
      'libs/cool-lib/src/lib/abc-two/abc-two.component.cy.ts',
      'should not overwrite abc-two'
    );
    tree.write(
      'libs/cool-lib/src/lib/abc-three/abc-three.component.cy.ts',
      'should not overwrite abc-three'
    );

    await cypressComponentConfiguration(tree, {
      project: 'cool-lib',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      name: 'abc',
      basePath: 'libs/cool-lib/src/lib',
    });

    expect(one.cy).toEqual('should not overwrite abc-one');
    expect(two.cy).toEqual('should not overwrite abc-two');
    expect(three.cy).toEqual('should not overwrite abc-three');
  });

  // TODO: should we support this?
  it.skip('should handle multiple components per file', async () => {
    await libraryGenerator(tree, {
      name: 'multiple-components',
      flat: true,
    });

    await componentGenerator(tree, {
      name: 'cmp-one',
      project: 'multiple-components',
      flat: true,
    });
    await componentGenerator(tree, {
      name: 'cmp-two',
      project: 'multiple-components',
      flat: true,
    });
    console.log(
      tree.read(
        'libs/multiple-components/src/lib/multiple-components.module.ts',
        'utf-8'
      )
    );
    tree.write(
      `libs/multiple-components/src/lib/cmp-one.component.ts`,
      `
import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'proj-cmp-one',
  templateUrl: './cmp-one.component.html',
  styleUrls: ['./cmp-one.component.css']
})
export class CmpOneComponent implements OnInit {

  constructor() { }

  ngOnInit(): void {
  }

}

@Component({
  selector: 'proj-cmp-one',
  template: '<h1>Hello World, {{abc}}</h1>',
  styles: []
})
export class CmpMultiComponent implements OnInit {
  @Input() name: string = 'abc'
  constructor() { }
  ngOnInit(): void {}
}
`
    );

    tree.write(
      '',
      `
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CmpOneComponent, CmpMultiComponent } from './cmp-one.component';
import { CmpTwoComponent } from './cmp-two.component';

@NgModule({
  imports: [
    CommonModule
  ],
  declarations: [
    CmpOneComponent,
    CmpTwoComponent
  ]
})
export class MultipleComponentsModule { }
`
    );

    await cypressComponentConfiguration(tree, {
      project: 'multiple-components',
      generateTests: true,
    });
    expect(
      tree.read(
        'libs/multiple-components/src/lib/cmp-one.component.cy.ts',
        'utf-8'
      )
    ).toEqual('');
  });
});

async function setup(
  tree: Tree,
  options: {
    name: string;
    project: string;
    standalone?: boolean;
    withInputs?: boolean;
    basePath?: string;
  }
) {
  for (const name of [
    `${options.name}-one`,
    `${options.name}-two`,
    `${options.name}-three`,
  ]) {
    await componentGenerator(tree, { project: options.project, name });

    if (options.withInputs) {
      const cmpPath = joinPathFragments(
        options.basePath,
        name,
        `${name}.component.ts`
      );
      const oldContent = tree.read(
        cmpPath,

        'utf-8'
      );

      const newContent = oldContent.replace(
        'constructor()',
        `
  @Input('buttonType') type = 'button';
  @Input() style: 'default' | 'fancy' | 'link' = 'default';
  @Input() age?: number;
  @Input() isOn = false;
  @Input() message: string | undefined;
  @Input() anotherProp: any;
  @Input() anotherNeverProp: never;

  constructor()`
      );

      tree.write(cmpPath, newContent);
    }
  }
}
function getCmpsFromTree(
  tree: Tree,
  options: { basePath: string; name: string }
) {
  return [
    `${options.name}-one`,
    `${options.name}-two`,
    `${options.name}-three`,
  ].map((n) => {
    expect(
      tree.exists(joinPathFragments(options.basePath, n, `${n}.component.ts`))
    ).toBeTruthy();
    expect(
      tree.exists(
        joinPathFragments(options.basePath, n, `${n}.component.cy.ts`)
      )
    ).toBeTruthy();
    return {
      cmp: tree.read(
        joinPathFragments(options.basePath, n, `${n}.component.ts`),
        'utf-8'
      ),
      cy: tree.read(
        joinPathFragments(options.basePath, n, `${n}.component.cy.ts`),
        'utf-8'
      ),
    };
  });
}
