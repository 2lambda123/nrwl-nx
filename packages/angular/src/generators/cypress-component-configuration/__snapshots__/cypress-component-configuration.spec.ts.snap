// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cypress Component Testing Configuration should work with complex component 1`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingOneComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex component 2`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingTwoComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex component 3`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingThreeComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 1`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingOneComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 2`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingTwoComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 3`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingThreeComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with secondary entry point libs 1`] = `
"import { MountConfig } from 'cypress/angular';
import { FancyButtonComponent } from './fancy-button.component';

describe(FancyButtonComponent.name, () => {
  const config: MountConfig<FancyButtonComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(FancyButtonComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with secondary entry point libs 2`] = `
"import { MountConfig } from 'cypress/angular';
import { StandaloneFancyButtonComponent } from './standalone-fancy-button.component';

describe(StandaloneFancyButtonComponent.name, () => {
  const config: MountConfig<StandaloneFancyButtonComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(StandaloneFancyButtonComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 1`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingOneComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 2`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingTwoComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 3`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingThreeComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 1`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingOneComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 2`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingTwoComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 3`] = `
"import { MountConfig } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     cy.mount(SomethingThreeComponent, config);
  })
})
"
`;
